### PR TITLE
Create note

### PR DESCRIPTION
**Pull request by sezaljain**

### CODE REVIEW

File: project.clj [hunk 1]
```diff
@@ -6,6 +6,7 @@
                  [thheller/shadow-cljs "2.8.45"]
                  [ragtime "0.8.0"]
                  [cljs-ajax "0.8.0"]
+                 [clj-time "0.15.2"]
```
Comment (kirang89) [inline project.clj:9]: > If you are using Java 8 or later, consider using the built-in Java Time instead of Joda Time -- and look at clojure.java-time if you want a Clojure wrapper for that

I'm assuming that you're using Java 8+, so this is something to keep in mind.

File: src/clj/nnts2/middleware.clj [hunk 2]
```diff
@@ -13,6 +13,10 @@
   [word]
   (-> word keyword str (str/replace "_" "-") (subs 1) keyword))
 
+(defn kebab->snake
```
Comment (kirang89) [inline src/clj/nnts2/middleware.clj]: Use https://github.com/clj-commons/camel-snake-kebab for doing these conversions.

File: src/clj/nnts2/middleware.clj [hunk 3]
```diff
@@ -65,16 +69,21 @@
...
-             (handler (assoc-in request [:session :user-info] user-info)))
-           (catch Exception e
-             (-> (res/response "Not authorized")
-                 (res/status 401)))))))
+    (let [access-token (or (get-in request [:session :ring.middleware.oauth2/access-tokens :google :token])
+                           (get-in request [:headers "authorization"]))]
+
+      (let [response (clj-http/get "https://www.googleapis.com/oauth2/v3/userinfo"
+                                   {:throw-exceptions false
+                                    :headers          {"Authorization" (if (not (str/starts-with? access-token "Bearer "))
```
Comment (kirang89) [inline src/clj/nnts2/middleware.clj:64]: It'd be cleaner to extract the token logic into the enclosing `let` and to call that here.

File: src/clj/nnts2/server.clj [hunk 4]
```diff
@@ -10,36 +10,43 @@
...
                                         status content-type]]))
 
+
 (defonce ^:private all-sessions (mem/memory-store))
 
+(def auth-routes (compojure.core/routes
+                   user/routes
+                   (context "/note" [] note/routes)))
+
 (defroutes app-routes
```
Comment (kirang89) [inline src/clj/nnts2/server.clj:30]: The convention is to move all routes to `routes.clj` and to simply define the middlewares and the server lifecycle methods here.

File: test/clj/nnts2/note/functional_test.clj [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+             [nnts2.note.handler :refer [create]]
+             [nnts2.note.db :refer [add]]))
+
+
+(use-fixtures :each fixtures/clear)
+(use-fixtures :once fixtures/setup fixtures/adduser)
+
+
+(deftest create-note-success
+  (testing "String title, content and valid user id -> note success"
```
Comment (kirang89) [inline test/clj/nnts2/note/functional_test.clj]: Consider writing simple, generic user level test strings rather than domain level ones, except for edge cases. For example, this could be converted to `should create note successfully for valid params`. This is only a convention of course, so feel free to ask around to get an idea of how folks write them.

**Comment by kirang89** [comment]
As a sidenote, it'd be pretty cool if you could post a gif illustrating how these actions work in the UI. This will also help reviewers get a feel of `nnts2` :).

See this PR as an example: https://github.com/nilenso/time-tracker-web-nxt/pull/45

Hope the review is useful :)

File: project.clj (hunk 1 above)
Comment (sezaljain) [reply project.clj:9]: got it. Was trying to use this library, but reverted those changes. Will add it via the built-in java time now

File: src/clj/nnts2/middleware.clj (hunk 2 above)
Comment (sezaljain) [reply src/clj/nnts2/middleware.clj]: ohh! this looks good, thanks!

File: src/clj/nnts2/server.clj (hunk 4 above)
Comment (sezaljain) [reply src/clj/nnts2/server.clj:30]: ok, we were thinking of separating routes requiring authentication from the ones  which wont.
And a few middleware are only used on the auth-routes. What will be your suggestion for that?

File: test/clj/nnts2/note/functional_test.clj (hunk 5 above)
Comment (sezaljain) [reply test/clj/nnts2/note/functional_test.clj]: Ok. should I be writing more tests here?

File: src/clj/nnts2/server.clj (hunk 4 above)
Comment (kirang89) [reply src/clj/nnts2/server.clj:30]: You could still keep them in `routes.clj`  as auth and non-auth routes and then combine them to form the final `routes`.

File: resources/migrations/003-create-note-table.edn [hunk 6]
```diff
@@ -0,0 +1,9 @@
+{:up ["CREATE TABLE notes (id serial primary key,
+                           title text,
+                           content text,
+                           created_by_id integer not null,
+                           created_at timestamptz not null default now(),
+                           updated_at timestamptz not null default now(),
+                           note_time timestamptz not null default now(),
+                           foreign key (created_by_id) references users (id));"]
```
Comment (scimetfoo) [inline resources/migrations/003-create-note-table.edn:8]: This can just say `created_by_id references users(id)` since you're not specifying a constraint name anyway.

File: src/clj/nnts2/db/note.clj [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+            [honeysql-postgres.helpers :as ph]
+            [clojure.java.jdbc :as jdbc]
+            [honeysql.core :as sql]
+            [nnts2.utils :as utils]
+            [nnts2.config :as config]))
+
+;; just change keywords from kebab to snake
+
+(defn clojure-data->sql-data [sub-form]
+  (into {} (for [[k v] sub-form]
```
Comment (scimetfoo) [inline src/clj/nnts2/db/note.clj]: Is there a reason to not do this in db middleware? I think it would be worth creating a db middleware.

File: src/clj/nnts2/model/user.clj [hunk 8]
```diff
@@ -1,4 +1,4 @@
-(ns nnts2.user.spec
+(ns nnts2.model.user
```
Comment (scimetfoo) [inline src/clj/nnts2/model/user.clj:1]: Worth calling this namespace `user-spec` since there could be other user models as well.

File: src/clj/nnts2/routes/note.clj [hunk 9]
```diff
@@ -0,0 +1,7 @@
+(ns nnts2.routes.note
+  (:require [compojure.core :refer [GET POST defroutes]]
+            [nnts2.handler.note :as handler]))
+
+(defroutes routes
+    (POST "/create" []
```
Comment (scimetfoo) [inline src/clj/nnts2/routes/note.clj:6]: The route could just be `"/note"`. Having the verb in the route makes it un-RESTful.

File: src/clj/nnts2/routes/user.clj [hunk 10]
```diff
@@ -0,0 +1,10 @@
+(ns nnts2.routes.user
+  (:require [compojure.core :refer [defroutes GET]]
+            [ring.util.response :as res]
+            [nnts2.handler.user :as user-handler]))
+
+(defroutes routes
+           (GET "/user" [] user-handler/create)
+
+           (GET "/user-info" []
```
Comment (scimetfoo) [inline src/clj/nnts2/routes/user.clj:9]: Unrelated: I think we should change this to say `"/google-user"` since we're starting to differentiate nnts user from google user. (This change doesn't need to be made in this PR. I was just pointing out.)

File: src/cljs/nnts2/events.cljs [hunk 11]
```diff
@@ -16,7 +19,7 @@
 
 
 (re-frame/reg-event-fx
-  ::get-user-info
+ ::get-user-info
   (fn [_ event]
     (user-api/get-info)))
```
Comment (scimetfoo) [inline src/cljs/nnts2/events.cljs:24]: This is a function with side effects. It's not "pure". Conventionally, with re-frame it's best to use [re-frame-http-fx](https://github.com/Day8/re-frame-http-fx) and have it return a map of `db` and `http-xhrio`. It also, to some degree, enforces better state management.

File: src/cljs/nnts2/note/events.cljs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+
+(re-frame/reg-event-fx
+ :note-submit
+ (fn [_ event]
+   (api/create-note (rest event))))
+
+(re-frame/reg-event-db
+ :note-get-list
+ (fn [db event]
+   db))
```
Comment (scimetfoo) [inline src/cljs/nnts2/note/events.cljs:22]: What's the purpose of this event? Seems unused. Also, it just returns the `db` as-is so there don't seem to be any effects registered for it. Let me know if this is intentional because what succeeds is not a part of this story.

File: src/cljs/nnts2/note/components.cljs [hunk 13]
```diff
@@ -0,0 +1,47 @@
+(ns nnts2.note.components
+  (:require [re-frame.core :as re-frame]
+            [reagent.core :as r]))
+
+(enable-console-print!)
+
+(def full (r/atom ""))
```
Comment (scimetfoo) [inline src/cljs/nnts2/note/components.cljs]: Should this be named something else? Might not be easy to gauge what this atom does.

File: src/cljs/nnts2/note/components.cljs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+(defn write-notes []
+  (let [content-text (r/atom "")
+        title-text (r/atom "")]
```
Comment (scimetfoo) [inline src/cljs/nnts2/note/components.cljs]: Maybe combine these and make an atomic map instead? Something like this:
````
[note (r/atom {:content ""
               :title   ""}]
````

File: test/clj/nnts2/fixtures.clj [hunk 15]
```diff
@@ -1,18 +1,36 @@
...
 (defn clear [test]
   (sql/with-db-transaction [db (config/db-spec)]
                            (sql/db-set-rollback-only! db)
                            (binding [config/db-spec (constantly db)]
                              (test))))
+
+
+
+(defn adduser
+  [tests]
```
Comment (scimetfoo) [inline test/clj/nnts2/fixtures.clj:30]: This seems like a helper more than a fixture. Can we not have this as a fixture? Might want to add a different user occasionally.

File: test/clj/nnts2/note/functional_test.clj (hunk 5 above)
Comment (scimetfoo) [reply test/clj/nnts2/note/functional_test.clj]: This seems to be testing both the note creation as well as the handler. Maybe separate those two out into different tests? Also, I'm following a naming convention similar to what Kiran mentioned. Let's stick to that?

File: src/cljs/nnts2/note/components.cljs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+(defn list-notes []
+  [:div [:h1 "I will list notes here"]])
+
+
+(defn note-panel []
+  [:div
+   ;[:div [list-notes]]
+   [:div [write-notes]]
+   [:div @full]])
```
Comment (scimetfoo) [inline src/cljs/nnts2/note/components.cljs]: What's the purpose of this? Seems redundant. Should this atom be removed?

File: test/clj/nnts2/note/functional_test.clj (hunk 5 above)
Comment (sezaljain) [reply test/clj/nnts2/note/functional_test.clj]: yes

File: src/clj/nnts2/routes/note.clj (hunk 9 above)
Comment (sezaljain) [reply src/clj/nnts2/routes/note.clj:6]: It currently clashes with a fronted route, will fix it in the next pr

File: src/clj/nnts2/routes/user.clj (hunk 10 above)
Comment (sezaljain) [reply src/clj/nnts2/routes/user.clj:9]: When we are asking use-info from frontend, as a client we dont care if the info is coming from google or nnts. In implementing this we have chosen to use google info as it is, but that should not affect the url

File: src/cljs/nnts2/note/events.cljs (hunk 12 above)
Comment (sezaljain) [reply src/cljs/nnts2/note/events.cljs:22]: This is an overlap bw this and next story. Wanted to hit this event on success of note-creation. Am calling note-submit-success event now but will fully integrate note-get-list in the next pr

File: src/cljs/nnts2/note/components.cljs (hunk 13 above)
Comment (sezaljain) [reply src/cljs/nnts2/note/components.cljs]: removed

File: src/clj/nnts2/routes/user.clj (hunk 10 above)
Comment (scimetfoo) [reply src/clj/nnts2/routes/user.clj:9]: Yes, that's true, but the response contains user info as structured by Google, which is different from how we persist it. Might be valuable to be explicit about that. The nnts user has different field names and id.

File: src/cljs/nnts2/note/components.cljs (hunk 14 above)
Comment (sezaljain) [reply src/cljs/nnts2/note/components.cljs]: This becomes a little complicated if you are trying to clear the form. I cant use what you have suggested in that case

File: test/clj/nnts2/fixtures.clj (hunk 15 above)
Comment (sezaljain) [reply test/clj/nnts2/fixtures.clj:30]: Make sense, I will move it in the future prs though. In this setup, I am totally using it as a fixture.
